### ISSUE TITLE
Selecting Text

### ISSUE THREAD

**Issue by tijsproost** [mattbryson/TouchSwipe-Jquery-Plugin#149]
Hi,

I am using the touchswipe plugin on my wrapper div.
Everything works normal except for selecting text inside the wrapper, i cant select anything inside my wrapper div.

Is there a fix for this issue or a work around?

**Comment by tridoan** [comment]
Also seeing this issue. Surprised there is no solution yet...?

**Comment by reardestani** [comment]
Hi, 

I had the same issue and I couldn't add `.noSwipe` class to any element. So I decide to detect the mobile device first then apply this plugin. It works well.

     // Code credit: http://stackoverflow.com/questions/3514784/what-is-the-best-way-to-detect-a-mobile-device-in-jquery
		function isMobile() {
		  try{ document.createEvent("TouchEvent"); return true; }
		  catch(e){ return false; }
		}

		if ( isMobile() == true ) {

    }

**Comment by lucazpf** [comment]
reardestani posted a cool workaround. For me works good

**Comment by davidbell** [comment]
That's what I ended up using, but surely a PC with a touch-sensitive screen would wrongly be identified as a mobile device just because it has the capability of touch, regardless of whether you were using a mouse at the time.

**Comment by spenat28** [comment]
While i am using swipe to open sidemenu, i cannot use *excludedElements* or *.noSwipe* ... dunno why this is not solved yet ... i thought, preventDefault = false would solve it, but it does not ...
Please, can You solve it? I have to use different lib right now for that, but i do like this one better.

**Comment by goesredy** [comment]
I solve this issue by attach swipe event only on touch. This is my code :
```
$(document).on("touchstart",function(event) {
	// Attach swipe event here.
});
```
So it also detect touchscreen in PC, rather than use mobile device detect.

**Comment by spenat28** [comment]
@goesredy but it's still a workaround, not solution of issue.

**Comment by goesredy** [comment]
@spenat28 Yep, it needs a hardwork

**Comment by mattbryson** [comment]
What exactly is the issue here - is this on touch devices? iOS, android, windows mobile?  or using a mouse?  

Using a mouse, clicking and dragging to select text is the same as swiping, so they are mutually exclusive.

I've just tested in iOS and long press to select text works along side swiping.

If someone provide a clear example of what the issue issue is I will have a look at it.

**Comment by lucazpf** [comment]
@mattbryson Using a mouse, on my site can't select the text. For me it's not a real bug, it needs only a workaround because ad You said dragging the mouse to select text it's like swaping

**Comment by mattbryson** [comment]
Yeah, this is a tricky one on mouse based input.

I the latest version, If you don’t “move” in an interaction, then it allows default events.  This means out the box things like buttons and a tags will trigger without the need to exclude them.

If your text is in a text input, the clicking in that input and then selecting may  work, but Ive not tested that scenario.

If you test is just part of a DIV, then there is no way for the plugin to know if the user should be selecting text or swiping.

m



-- 
Matt Bryson


On 29 April 2016 at 19:35:12, Luca Scalvi (notifications@github.com) wrote:

@mattbryson Using a mouse, on my site can't select the text. For me it's not a real bug, it needs only a workaround because ad You said dragging the mouse to select text it's like swaping

**Comment by mrclMD** [comment]
For me its not possible to select text on a desktop pc, too. I can live with it. But I have another problem. My select boxes are not opening! This is a bug for me with which I cannot live, so I decided to just enable swipe on touch devices with the function @reardestani provided above.

**Comment by dev-101** [comment]
Just discovered the library (great work! :)) and immediately noticed text selection issue with mouse-input devices. At first, I have bind listener to the entire document (body) per my requirements, but soon realized this is a bad idea, because of non-functional text selection issue.

One way is to use above mentioned mobile detection (3rd party library with high accuracy is recommended) and enable it only on touch supported devices, where this problem with text selection does not exist since long-touch initiates text selection mode on OS level.

Or to limit swipe to an element that does not contain any text (e.g. top bar or similar). A compromise for the time being.

**Comment by thefraj** [comment]
Does anyone have a reliable solution to this? This is quite a critical one which baiscally means we can't use this plugin in production at all, but it is a great plugin and does add the neat swipe functionality.

**Comment by mattbryson** [comment]
@thefraj what is your usecase - you want swiping on touch interaction, but text selection on mouse interaction?

**Comment by thefraj** [comment]
Hi @mattbryson  thanks - basically for mobile we'd like to be able to handle swipe gestures for mobile navigation, but using thus plugin bascially stops all regular touches (including text highligthing) from working, which is our sticking point. For this site, it's important users can highlight text

I'm aware we can add .noSwipe class, but for a few reasons this isn't ideal for us.  We then tried the trick of using 'IsMobile' check (creating touch event, testing success of event for deciding whether or not to involke touch) while this doesn't give any errors, it also appears to stop swipe working altogether, below is the sort of code we are using for this:

function isMobile() {
  try{ document.createEvent("TouchEvent"); return true; }
  catch(e){ return false; }
}

if ( isMobile() == true ) {
	var code = '<script>$("body").swipe({allowPageScroll:"vertical",swipe:function(e,o){if(vpWidth<768){var h=$(this).hasClass(n),s=$("#slidemenu"),c=$("#navbar-height-col"),l=$("#page-content, .navbar, .navbar-header, .navbar-toggle "),a=$(this),v=$("html");if(o==="up"||o==="down")return!1;o!=="right"||h?o==="left"&&h&&(v.removeAttr("style"),s.stop().animate({left:u}),c.stop().animate({left:f}),$(i).stop().animate({left:0}),$(r).stop().animate({left:0}),a.toggleClass(n),s.toggleClass(n),l.toggleClass(n)):(v.css({overflow:"hidden",width:"100%"}),s.stop().animate({left:0}),c.stop().animate({left:0}),$(i).stop().animate({left:t}),$(r).stop().animate({left:t}),a.toggleClass(n),s.toggleClass(n),l.toggleClass(n))}}});<\/script>';		
		
	$('body').append(code);		
	
}
			else {
				alert("touch not enabled");
			}
		</script>	

The mobile check is working because we get the appropriate alert, however, the swipe code doesn't behave when loaded this way.

Basically if there is a way to disable this on desktops that would be perfect

**Comment by katerlouis** [comment]
I used Ben Majors ["jquerytouchevents"](https://github.com/benmajor/jQuery-Touch-Events) before I switched to this solution here, due to errors in conjunction with other libraries etc. Bens solution didn't block text selection for mouse usage. 

I'm curious why Matts solution goes that route? There must be a reason why this hasn't changed in more than 5 years.

**Comment by mecheverria** [comment]
Until now I used version 1.6.6 and it didn't give me problems. But it was changed to version 1.6.18 and the imput-select stopped working. So for now I continue with 1.6.6 despite other problems it has.

**Comment by mecheverria** [comment]
Until now I used version 1.6.6 and it didn't give me problems. But it was changed to version 1.6.18 and the imput-select stopped working. So for now I continue with 1.6.6 despite other problems it has.
